### DISCUSSION TITLE
GeCos-W-Sens - Environment Sensor - BME680 & APDS-9960 - Custom Build - need Help

### DISCUSSION THREAD

**Discussion by elbman** [arendst/Tasmota#13648] [Q&A] [answered]
I have just flashed an interesting environment sensor with Tasmota: 

GeCos-W-Sens - not listed under blackadders Tasmota Supported Devices Repository yet
https://www.gedad.de/shop/gecos-wireless/#cc-m-product-14884828622

Buildin BME680 - works fine with tasmota-sensors.bin but
there is also an APDS-9960 connected to I2C and regognized by i2cscan

Concerning the light sensor: "This feature is not included in precompiled binaries - To use it you must compile your build."

Is there someone willing to help me and build a binary for me and others, supporting BME680 & APDS-9960?

I don't trust myself to do that

**Comment by ascillato** [comment, marked as answer]
> Is there someone willing to help me and build a binary for me and others, supporting BME680 & APDS-9960?

Hi, it is super easy. Click in the following link and choose what you want and Tasmocompiler will compile and give you that firmware.

https://gitpod.io/#https://github.com/benzino77/tasmocompiler

Instructions on how to use it: https://github.com/benzino77/tasmocompiler

**Reply by elbman** [reply]
Thank you, i did it !

**Reply by peterrn2** [reply]
I bought one GeCoS-W-Sens and I compiled Tasmota successfully with VSCode by decommenting the lines concerning BME680 and APDS9660,
Firmware did upload and by assigning GPIO2 to SDA and GPIO14 to SCL I2Cscan prompts {"I2CScan":"Device(s) found at 0x39 0x76"}.
Using Sensor27 0 switches on the output of APDS9660 in the webui. But BME680 does not show.

Is there anybody out there who can give me hints on using the GeCoS-W-Sens with Tasmota?
Thanks, Peter.

**Reply by sfromis** [reply]
Better verify that BME680 support is actually included in your binary. You can do that by looking for `USE_BME680` in the output of the [decode-status](https://github.com/tasmota/decode-status) tool, section Features.

**Reply by barbudor** [reply]
With the command `I2CDriver` you will be able to check which I2C drivers is loaded in your binary and which are enabled.
Numbers with a `!` in front are disabled driver
Numbers without the `!` are enabled.
You can enable a driver with the command `I2CDriverXX 1` or disable with `I2CDriverXX 0` where `XX` is the number of the driver (no space between the command and the number.
Note that Tasmota will reboot after the command. If you want to send multiple I2CDriver commands, use `Backlog` with all commands separated by `;` so Tasmota will reboot only once.

The table with the I2CDrivers number is here : https://tasmota.github.io/docs/I2CDEVICES/
You will see that the BMP driver is number 10

It is a good practice to disable all drivers that you don't need to avoid any conflict.
So in your case, only 10 and 21 should be enabled, all others should have the `!`.

**Reply by sfromis** [reply]
Using I2CDriver does not work well for BME680, as this is in the regular BMP driver number 10. But that driver does not support BME680, unless the build includes USE_BME680. That's why I suggested using decode-status, instead of the normally preferable I2CDriver :-)

**Reply by sfromis** [reply]
When it comes to disabling unneeded drivers, when I do a build (like to add drivers) I normally include
```
#define I2CDRIVERS_0_31 0x00000000
#define I2CDRIVERS_32_63 0x00000000
#define I2CDRIVERS_64_95 0x00000000
```
This will disable all I2C drivers, and then I can use the command I2CDriver to activate the ones I want to use. Of course, I could also count bits in the masks to specifically configure, but that would not be a good fit for me using the same build for different sensor mixes.

**Reply by peterrn2** [reply]
Thanks for your fast replies.

The´decode-status tool will not work on my Windows-PC.
I tried the executables from build 84 and 83 and the 32bit-version from build 84.
These executables finish saying 
Traceback (most recent call last):
  File "C:\Users\Peter\OneDrive\Dokumente\VSCode\decode-status-master\decode-status.py", line 45, in <module> import requests
ModuleNotFoundError: No module named 'requests'
So I got stuck with this tool.

My binary includes:
```
// -- I2C sensors ---------------------------------
#define USE_I2C                                  // I2C using library wire (+10k code, 0k2 mem, 124 iram)
#define I2CDRIVERS_0_31        0xFFFFFFFF          // Enable I2CDriver0  to I2CDriver31
#define I2CDRIVERS_32_63       0xFFFFFFFF          // Enable I2CDriver32 to I2CDriver63
#define I2CDRIVERS_64_95       0xFFFFFFFF          // Enable I2CDriver64 to I2CDriver95

#define USE_BME680                           // Enable support for BME680 sensor using Bosch BME680 library (+4k code)

#define USE_APDS9960                           // [I2cDriver21] Enable APDS9960 Proximity Sensor (I2C address 0x39). Disables SHT and VEML6070 (+4k7 code)
    #define USE_APDS9960_GESTURE                   // Enable APDS9960 Gesture feature (+2k code)
    #define USE_APDS9960_PROXIMITY                 // Enable APDS9960 Proximity feature (>50 code)
    #define USE_APDS9960_COLOR                     // Enable APDS9960 Color feature (+0.8k code)
    #define USE_APDS9960_STARTMODE  0              // Default to enable Gesture mode
```

Due to few spare time it will take some days for me to follow your hints enablind/disabling drivers by I2CDriver.
I will report my next steps and results.

**Reply by barbudor** [reply]
There seems to be a problem with the prebuilt exe for decode-status. I can't even built a working version myself.
If you have python installed, you can try to run the `decode-status.py` directly with `python decode-status.py -d ip.of.tasmota`
You can also provide us the output of `Status 4` command and we will decrypt the data for you.

**Reply by peterrn2** [reply]
Well I reloaded decode-status from the repositry, executed 
python decode-status.py -d 192.... 
and it throws the same error. Here is the output of Status 4:
`{"StatusMEM":{"ProgramSize":619,"Free":384,"Heap":25,"ProgramFlashSize":4096,"FlashSize":4096,"FlashChipId":"164020","FlashFrequency":40,"FlashMode":3,"Features":["00000809","8FDAC787","04368001","400000CF","000013C0","C000F981","00004004","00001000","00000020"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,16,18,19,20,21,22,24,26,27,29,30,35,37,45","Sensors":"1,2,3,4,5,6,27"}}
`
Thank you for your offer to decrypt the data for me.

**Reply by barbudor** [reply]
From code in support_features.ino :
```
#ifdef USE_I2C
    feature1 |= 0x00000004;  // tasmota.ino
#endif
...
#if defined(USE_I2C) && defined(USE_BMP)
    feature3 |= 0x00000400;  // xsns_09_bmp.ino
#endif
#if defined(USE_I2C) && defined(USE_BMP) && defined(USE_BME680)
    feature3 |= 0x00000800;  // xsns_09_bmp.ino - BME680
#endif
```
and yours
```
"Features":["00000809","8FDAC787","04368001","400000CF","000013C0","C000F981","00004004","00001000","00000020"]
```
Feature1 = "8FDAC787" I2C included
Feature3 = "400000CF" obviously, niether USE_BMP nor USE_BME280 are there

So you have a build issue
Are you building `tasmota` ?
Beware that variants such as `tasmota-sensors` cannot be customized as the default configuration is likely to override your override
Always build `tasmota`

**Reply by sfromis** [reply]
Ran my status4 decoder. USE_BME680 not set, would have been bit 2**11 in 4rd features word, code 3-11 in this table of included features.
```
1-0  USE_ENERGY_MARGIN_DETECTION 1
1-1  USE_LIGHT                   1
1-2  USE_I2C                     1
1-7  USE_WEBSERVER               1
1-8  WEBSERVER_ADVERTISE         1
1-9  USE_EMULATION_HUE           1
1-10 MQTT_PUBSUBCLIENT           1
1-14 USE_ARILUX_RF               1
1-15 USE_WS2812                  1
1-17 USE_IR_REMOTE               1
1-19 USE_IR_RECEIVE              1
1-20 USE_DOMOTICZ                1
1-22 USE_HOME_ASSISTANT          1
1-23 USE_SERIAL_BRIDGE           1
1-24 USE_TIMERS                  1
1-25 USE_SUNRISE                 1
1-26 USE_TIMERS_WEB              1
1-27 USE_RULES                   1
1-31 USE_ENERGY_POWER_LIMIT      1
2-0  USE_CONFIG_OVERRIDE         1
2-15 USE_TUYA_DIMMER             1
2-17 USE_ARMTRONIX_DIMMERS       1
2-18 USE_SM16716                 1
2-20 USE_EMULATION_WEMO          1
2-21 USE_SONOFF_IFAN             1
2-26 VTABLES_IN_FLASH            1
3-0  USE_COUNTER                 1
3-1  USE_ADC_VCC                 1
3-2  USE_ENERGY_SENSOR           1
3-3  USE_PZEM004T                1
3-6  USE_DS18x20                 1
3-7  USE_DHT                     1
3-30 USE_APDS9960                1
4-6  USE_HLW8012                 1
4-7  USE_CSE7766                 1
4-8  USE_MCP39F501               1
4-9  USE_PZEM_AC                 1
4-12 USE_PZEM_DC                 1
5-0  USE_BUZZER                  1
5-7  USE_SM2135                  1
5-8  USE_SHUTTER                 1
5-11 USE_DEEPSLEEP               1
5-12 USE_SONOFF_SC               1
5-13 USE_SONOFF_RF               1
5-14 USE_SONOFF_L1               1
5-15 USE_EXS_DIMMER              1
5-30 USE_DEVICE_GROUPS           1
5-31 USE_PWM_DIMMER              1
6-2  USE_SONOFF_D1               1
6-14 USE_BL09XX                  1
7-12 USE_SHELLY_DIMMER           1
8-5  USE_ENERGY_DUMMY            1
```

**Reply by peterrn2** [reply]
Thank you! And it works! I see outputs of both sensors in the Tasmota-webui.
The Solution was barbudors info that BME-driver was not included in my bin.
Also helpful to know that tasmota-sensors is not a good base to start with non-standard-sensors.
So yesterday I tried to generate a firmware via Gitpod.
Today I activated "#define USE_BME" - until now this was out of my mind.
Directly BME680s output came up. With "Sensor27 0" APDS9960 joined the webui.
So now I am going to check MQTT information and work on HomeAssistand.
Thanks again, great work and helpf from you!
